### PR TITLE
Ticket2397 emma fermi chopper

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
### Description of work

Add an IOC for the EMMA fermi chopper, which is an SKF MB350 PC.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/2397

### Acceptance criteria

Note: Some of these items _could_ be tested in a limited way on the real hardware if the reviewer thinks this is absolutely necessary. However, do not attempt to move the chopper.

- [x] Can read back actual frequency
- [x] Can read back phase setpoint
- [x] Can read back actual phase
- [x] Can read back phase accuracy window
- [x] Can read back phase % ok
- [x] Can read back rotator angle
- [x] Can read back statuses
- [x] Can read back motor direction
- [x] Can read back interlocks
- [x] Can set frequency
- [x] Can set phase
- [x] Can set phase window
- [x] Can start the chopper
- [x] Can stop the chopper
- [x] All setpoints get autosaved, as per labview driver.

---

#### Code Review

- [ ] **Pertitent information and a copy of the manual has been stored in the [wiki](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Specific-Device-IOC)**
- [ ] Does the IOC conform to IBEX standards?
    - [PV naming](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/PV-Naming).
    - [Disable records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Disable-records)
    - [Record simulation](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation)
    - [Finishing touches](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Finishing-Touches)
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)?
- [ ] Have the changes been recorded appropriately in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Functional Tests

- IOC responds correctly in:
    - [ ] Devsim mode
    - [ ] Recsim mode
    - [ ] Real device, if available
- [ ] Supplementary IOCs (`..._0n` where `n>1`) run correctly
- [ ] Log files do not report undefined macros (serach for `macLib: macro` to find instances of `macLib: macro [macro name] is undefined...`

### Final steps

- [ ] Update the IOC submodule in the main EPICS repo. See [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by John-Holt-Tessella** [comment]
- [ ] Add manual and info to https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Specific-Device-IOC

File: SKFMB350/SKFMB350-IOC-01App/Db/chopper.db [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    field(VAL, "$(PARKPOS)")
+    field(ZNAM, "CLOSED")
+    field(ONAM, "OPEN")
+    field(PINI, "YES")
+}
+
+record(calcout, "$(P)ANGLE:PARK:OPEN:CALC") {
+    field(INPA, "$(P)ANGLE:PARK:OPEN CP MS")
+    field(INPB, "$(P)ANGLE:PARK:SELECTOR CP MS")
+    field(CALC, "B=1?A:0")
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply SKFMB350/SKFMB350-IOC-01App/Db/chopper.db]: I tried this but you get a race between PARK:OPEN and PARK:SELECTOR initializing, so sometimes you get two setpoints set in quick succession

File: SKFMB350/SKFMB350-IOC-01App/Db/chopper.db [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+record(ao, "$(P)ANGLE:PARK:CLOSED") {
+    field(VAL, "$(PARKCLOSED)")
+    field(PINI, "YES")
+}
+
+record(bo, "$(P)ANGLE:PARK:SELECTOR") {
+    field(VAL, "$(PARKPOS)")
+    field(ZNAM, "CLOSED")
+    field(ONAM, "OPEN")
+    field(PINI, "YES")
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply SKFMB350/SKFMB350-IOC-01App/Db/chopper.db]: Yes this is what the labview does currently.

File: SKFMB350/SKFMB350-IOC-01App/Db/chopper.db [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+alias("$(P)SIM:FREQ:SP", "$(P)SIM:FREQ")
+
+
+#
+# Phase
+#
+
+record(ai, "$(P)PHAS") {
+    field(EGU, "us")
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply SKFMB350/SKFMB350-IOC-01App/Db/chopper.db]: Yes

File: SKFMB350/SKFMB350-IOC-01App/Db/status.template [hunk 4]
```diff
@@ -0,0 +1,16 @@
+record(bi, "$(P)STAT:$(SNAME)") {
+	field(DESC, "Status: $(SNAME)")
+	field(INP, "$(P)STAT.$(FIELD) CP MS")
+	field(ZNAM, "$(ZNAM)")
+	field(ONAM, "$(ONAM)")
+	info(archive, "VAL")
+	info(interest, "MEDIUM")
+	field(SIML, "$(P)SIM")
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply SKFMB350/SKFMB350-IOC-01App/Db/status.template]: I think this it's hard to tell which of these statuses are actually alarms and which ones are valid conditions that the chopper can be in.

In general my feeling it to put alarms on the interlocks as these are clearly "bad" conditions but to not put them on statuses.

**Comment by Tom-Willemsen** [comment]
Additional documentation added to https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/SKF_MB350_Chopper

Manual is already on the shared drive.

**Pull request merged**